src/allmydata/storage/http_server.py

```diff
@@ -26,38 +30,83 @@
 from .http_client import swissnum_auth_header
 
 
-def _authorization_decorator(f):
+class Secrets(Enum):
+    """Different kinds of secrets the client may send."""
+    LEASE_RENEW = "lease-renew-secret"
+    LEASE_CANCEL = "lease-cancel-secret"
+    UPLOAD = "upload-secret"
+
+
+class ClientSecretsException(Exception):
+    """The client did not send the appropriate secrets."""
+
+
+def _extract_secrets(header_values, required_secrets):  # type: (List[str], Set[Secrets]) -> Dict[Secrets, bytes]
+    """
+    Given list of values of ``X-Tahoe-Authorization`` headers, and required
+    secrets, return dictionary mapping secrets to decoded values.
+
+    If too few secrets were given, or too many, a ``ClientSecretsException`` is
+    raised.
+    """
+    string_key_to_enum = {e.value: e for e in Secrets}
+    result = {}
+    try:
+        for header_value in header_values:
+            string_key, string_value = header_value.strip().split(" ", 1)
+            key = string_key_to_enum[string_key]
+            value = b64decode(string_value)
+            if key in (Secrets.LEASE_CANCEL, Secrets.LEASE_RENEW) and len(value) != 32:
+                raise ClientSecretsException("Lease secrets must be 32 bytes long")
+            result[key] = value
+    except (ValueError, KeyError):
+        raise ClientSecretsException("Bad header value(s): {}".format(header_values))
+    if result.keys() != required_secrets:
+        raise ClientSecretsException(
+            "Expected {} secrets, got {}".format(required_secrets, result.keys())
+        )
+    return result
+
+
+def _authorization_decorator(required_secrets):
     """
     Check the ``Authorization`` header, and (TODO: in later revision of code)
     extract ``X-Tahoe-Authorization`` headers and pass them in.
     """
+    def decorator(f):
+        @wraps(f)
+        def route(self, request, *args, **kwargs):
+            if request.requestHeaders.getRawHeaders("Authorization", [None])[0] != str(
```

Comment:
this should probably use `timing_safe_compare` (or however we call it here) I think, or clients may be able to learn the correct swissnum.